okhttp/src/jvmTest/kotlin/okhttp3/TrailersTest.kt

```diff
@@ -0,0 +1,361 @@
+    this.server = server
+  }
+
+  @Test
+  fun trailersHttp1() {
+    trailers(Protocol.HTTP_1_1)
+  }
+
+  @Test
+  fun trailersHttp2() {
+    trailers(Protocol.HTTP_2)
+  }
+
+  private fun trailers(protocol: Protocol) {
+    enableProtocol(protocol)
+
+    server.enqueue(
+      MockResponse
+        .Builder()
+        .addHeader("h1", "v1")
+        .trailers(headersOf("t1", "v2"))
+        .body(protocol, "Hello")
+        .build(),
+    )
+
+    val call = client.newCall(Request(server.url("/")))
+    call.execute().use { response ->
+      val source = response.body.source()
+      assertThat(response.header("h1")).isEqualTo("v1")
+      assertThat(source.readUtf8()).isEqualTo("Hello")
+      assertThat(response.trailers()).isEqualTo(headersOf("t1", "v2"))
+      assertThat(response.trailers()).isEqualTo(headersOf("t1", "v2")) // Idempotent.
+    }
+  }
+
+  @Test
+  fun trailersWithoutReadingFullResponseBodyHttp1() {
+    trailersWithoutReadingFullResponseBody(Protocol.HTTP_1_1)
+  }
+
+  @Test
+  fun trailersWithoutReadingFullResponseBodyHttp2() {
+    trailersWithoutReadingFullResponseBody(Protocol.HTTP_2)
+  }
+
+  private fun trailersWithoutReadingFullResponseBody(protocol: Protocol) {
+    enableProtocol(protocol)
+
+    server.enqueue(
+      MockResponse
+        .Builder()
+        .addHeader("h1", "v1")
+        .trailers(headersOf("t1", "v2"))
+        .body(protocol, "Hello")
+        .build(),
+    )
+
+    val call = client.newCall(Request(server.url("/")))
+    call.execute().use { response ->
+      assertThat(response.header("h1")).isEqualTo("v1")
+      assertThat(response.trailers()).isEqualTo(headersOf("t1", "v2"))
+    }
+  }
+
+  @Test
+  @Disabled
+  fun trailersAndCacheHttp1() {
+    trailersAndCache(Protocol.HTTP_1_1)
+  }
+
+  @Test
+  @Disabled
+  fun trailersAndCacheHttp2() {
+    trailersAndCache(Protocol.HTTP_2)
+  }
+
+  private fun trailersAndCache(protocol: Protocol) {
+    enableProtocol(protocol)
+
+    server.enqueue(
+      MockResponse
+        .Builder()
+        .addHeader("h1", "v1")
+        .addHeader("Cache-Control: max-age=30")
+        .body(protocol, "Hello")
+        .trailers(headersOf("t1", "v2"))
+        .build(),
+    )
+
+    val call1 = client.newCall(Request(server.url("/")))
+    call1.execute().use { response ->
+      val source = response.body.source()
+      assertThat(response.header("h1")).isEqualTo("v1")
+      assertThat(source.readUtf8()).isEqualTo("Hello")
+      assertThat(response.trailers()).isEqualTo(headersOf("t1", "v2"))
+    }
+
+    val call2 = client.newCall(Request(server.url("/")))
+    call2.execute().use { response ->
+      val source = response.body.source()
+      assertThat(response.header("h1")).isEqualTo("v1")
+      assertThat(source.readUtf8()).isEqualTo("Hello")
+      assertThat(response.trailers()).isEqualTo(headersOf("t1", "v2"))
+    }
+  }
+
+  @Test
+  fun delayBeforeTrailersHttp1() {
+    delayBeforeTrailers(Protocol.HTTP_1_1)
```

Comment:
This manual no-burst stuff is the worst. I haven’t done it here 'cause I’m afraid of how our MockWebServer JUnit 5 thing will interact with Burst.

Reply:
(More precisely, Burst is broken when you do fancy things with JUnit 5)

Reply:
I don’t know if we can fix it anywhere. Burst kind of assumes it can do its own naughty business with parameters, and it might not have enough information to leave the JUnit-5 managed ones alone.